File: ios/Saga/Sources/Models/Social/Availability.swift
import Foundation

// MARK: - Availability

/// A user's posted availability for hangouts
struct Availability: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let userId: String
    let hangoutType: HangoutType
    let title: String?
    let description: String?
    let startTime: Date
    let endTime: Date
    let location: String?
    let locationLat: Double?
    let locationLng: Double?
    let radiusKm: Double?
    let status: AvailabilityStatus
    let createdOn: Date

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case hangoutType = "hangout_type"
        case title
        case description
        case startTime = "start_time"
        case endTime = "end_time"
        case location
        case locationLat = "location_lat"
        case locationLng = "location_lng"
        case radiusKm = "radius_km"
        case status
        case createdOn = "created_on"
    }

    /// Whether the availability is currently active
    var isActive: Bool {
        status == .active && endTime > Date()
    }

    /// Time range display
    var timeRange: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .short
        return "\(formatter.string(from: startTime)) - \(formatter.string(from: endTime))"
    }
}

// MARK: - Hangout Type

enum HangoutType: String, Codable, Sendable, CaseIterable {
    case talkItOut = "talk_it_out"
    case hereToListen = "here_to_listen"
    case concreteActivity = "concrete_activity"
    case mutualInterest = "mutual_interest"
    case meetAnyone = "meet_anyone"

    var displayName: String {
        switch self {
        case .talkItOut: return "Talk It Out"
        case .hereToListen: return "Here to Listen"
        case .concreteActivity: return "Concrete Activity"
        case .mutualInterest: return "Mutual Interest"
        case .meetAnyone: return "Meet Anyone"
        }
    }

    var description: String {
        switch self {
        case .talkItOut: return "I want to talk about something on my mind"
        case .hereToListen: return "I'm available to listen and support"
        case .concreteActivity: return "I have a specific activity in mind"
        case .mutualInterest: return "Looking to connect over shared interests"
        case .meetAnyone: return "Open to meeting new people"
        }
    }

    var iconName: String {
        switch self {
        case .talkItOut: return "bubble.left.and.bubble.right"
        case .hereToListen: return "ear"
        case .concreteActivity: return "figure.walk"
        case .mutualInterest: return "heart.circle"
        case .meetAnyone: return "person.wave.2"
        }
    }
}

// MARK: - Availability Status

enum AvailabilityStatus: String, Codable, Sendable, CaseIterable {
    case active
    case matched
    case expired
    case cancelled

    var displayName: String {
        switch self {
        case .active: return "Active"
        case .matched: return "Matched"
        case .expired: return "Expired"
        case .cancelled: return "Cancelled"
        }
    }
}

// MARK: - Nearby Availability

/// An availability found nearby with user info
struct NearbyAvailability: Codable, Sendable {
    let availability: Availability
    let user: PublicProfile
    let distanceKm: Double
    let compatibilityScore: Double?

    enum CodingKeys: String, CodingKey {
        case availability
        case user
        case distanceKm = "distance_km"
        case compatibilityScore = "compatibility_score"
    }
}

// MARK: - Hangout Request

/// A request to hang out based on an availability posting
struct HangoutRequest: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let availabilityId: String
    let requesterId: String
    let message: String?
    let status: HangoutRequestStatus
    let createdOn: Date
    let respondedOn: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case availabilityId = "availability_id"
        case requesterId = "requester_id"
        case message
        case status
        case createdOn = "created_on"
        case respondedOn = "responded_on"
    }
}

// MARK: - Hangout Request Status

enum HangoutRequestStatus: String, Codable, Sendable, CaseIterable {
    case pending
    case accepted
    case declined
    case cancelled

    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .accepted: return "Accepted"
        case .declined: return "Declined"
        case .cancelled: return "Cancelled"
        }
    }

    var iconName: String {
        switch self {
        case .pending: return "clock"
        case .accepted: return "checkmark.circle.fill"
        case .declined: return "xmark.circle.fill"
        case .cancelled: return "minus.circle.fill"
        }
    }
}

// MARK: - Hangout Request Display

/// A hangout request with full user and availability info
struct HangoutRequestDisplay: Codable, Sendable {
    let request: HangoutRequest
    let requester: PublicProfile
    let availability: Availability
}

// MARK: - Hangout

/// A confirmed hangout between users
struct Hangout: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let availabilityId: String
    let participants: [String]
    let scheduledTime: Date?
    let location: String?
    let status: HangoutStatus
    let createdOn: Date

    enum CodingKeys: String, CodingKey {
        case id
        case availabilityId = "availability_id"
        case participants
        case scheduledTime = "scheduled_time"
        case location
        case status
        case createdOn = "created_on"
    }
}

// MARK: - Hangout Status

enum HangoutStatus: String, Codable, Sendable, CaseIterable {
    case scheduled
    case completed
    case cancelled
    case noShow = "no_show"

    var displayName: String {
        switch self {
        case .scheduled: return "Scheduled"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        case .noShow: return "No Show"
        }
    }
}

// MARK: - Request Types

struct CreateAvailabilityRequest: Codable, Sendable {
    let hangoutType: HangoutType
    let title: String?
    let description: String?
    let startTime: Date
    let endTime: Date
    let location: String?
    let locationLat: Double?
    let locationLng: Double?
    let radiusKm: Double?

    enum CodingKeys: String, CodingKey {
        case hangoutType = "hangout_type"
        case title
        case description
        case startTime = "start_time"
        case endTime = "end_time"
        case location
        case locationLat = "location_lat"
        case locationLng = "location_lng"
        case radiusKm = "radius_km"
    }
}

struct UpdateAvailabilityRequest: Codable, Sendable {
    var title: String?
    var description: String?
    var startTime: Date?
    var endTime: Date?
    var location: String?
    var locationLat: Double?
    var locationLng: Double?
    var radiusKm: Double?
    var status: AvailabilityStatus?

    enum CodingKeys: String, CodingKey {
        case title
        case description
        case startTime = "start_time"
        case endTime = "end_time"
        case location
        case locationLat = "location_lat"
        case locationLng = "location_lng"
        case radiusKm = "radius_km"
        case status
    }
}

struct CreateHangoutRequestRequest: Codable, Sendable {
    let message: String?
}

struct RespondHangoutRequest: Codable, Sendable {
    let accept: Bool
    let message: String?
}

File: api/internal/model/availability.go
package model

import "time"

// HangoutType represents the type of hangout being sought
type HangoutType string

const (
	HangoutTypeTalkItOut        HangoutType = "talk_it_out"       // Need a listening ear
	HangoutTypeHereToListen     HangoutType = "here_to_listen"    // Ready to support someone
	HangoutTypeConcreteActivity HangoutType = "concrete_activity" // Specific activity/venue
	HangoutTypeMutualInterest   HangoutType = "mutual_interest"   // Connect over shared interest
	HangoutTypeMeetAnyone       HangoutType = "meet_anyone"       // Open to anyone
)

// HangoutTypeInfo provides display information for hangout types
type HangoutTypeInfo struct {
	Type        HangoutType `json:"type"`
	Label       string      `json:"label"`
	Description string      `json:"description"`
	Icon        string      `json:"icon"`
}

// GetHangoutTypeInfo returns display info for each type
func GetHangoutTypeInfo() []HangoutTypeInfo {
	return []HangoutTypeInfo{
		{
			Type:        HangoutTypeTalkItOut,
			Label:       "Talk It Out",
			Description: "I have something on my mind I'd like to talk through",
			Icon:        "bubble.left.fill",
		},
		{
			Type:        HangoutTypeHereToListen,
			Label:       "Here to Listen",
			Description: "I'm in a good headspace and happy to listen",
			Icon:        "ear.fill",
		},
		{
			Type:        HangoutTypeConcreteActivity,
			Label:       "Concrete Activity",
			Description: "I want to do a specific thing with someone",
			Icon:        "figure.walk",
		},
		{
			Type:        HangoutTypeMutualInterest,
			Label:       "Mutual Interest",
			Description: "I want to connect with someone who shares an interest",
			Icon:        "heart.fill",
		},
		{
			Type:        HangoutTypeMeetAnyone,
			Label:       "Meet Anyone",
			Description: "I'm open to meeting new people, surprise me",
			Icon:        "sparkles",
		},
	}
}

// AvailabilityStatus constants
type AvailabilityStatus string

const (
	AvailabilityStatusAvailable AvailabilityStatus = "available"
	AvailabilityStatusMaybe     AvailabilityStatus = "maybe"
	AvailabilityStatusBusy      AvailabilityStatus = "busy"
)

// AvailabilityLocation represents internal location with radius for matching
type AvailabilityLocation struct {
	Lat    float64 `json:"lat"`
	Lng    float64 `json:"lng"`
	Radius float64 `json:"radius"` // in km
}

// HangoutLocation represents a hangout location
type HangoutLocation struct {
	Lat   float64 `json:"lat"`
	Lng   float64 `json:"lng"`
	Venue *string `json:"venue,omitempty"`
}

// Availability represents a user's availability window
type Availability struct {
	ID                  string                `json:"id"`
	UserID              string                `json:"user_id"`
	Status              AvailabilityStatus    `json:"status"`
	StartTime           time.Time             `json:"start_time"`
	EndTime             time.Time             `json:"end_time"`
	Location            *AvailabilityLocation `json:"-"` // Never expose exact location
	HangoutType         HangoutType           `json:"hangout_type"`
	ActivityDescription *string               `json:"activity_description,omitempty"` // For concrete_activity
	ActivityVenue       *string               `json:"activity_venue,omitempty"`       // Specific venue
	InterestID          *string               `json:"interest_id,omitempty"`          // For mutual_interest
	MaxPeople           int                   `json:"max_people"`
	Note                *string               `json:"note,omitempty"`
	Visibility          string                `json:"visibility"` // circles, public
	ExpiresAt           time.Time             `json:"expires_at"`
	CreatedOn           time.Time             `json:"created_on"`
	UpdatedOn           time.Time             `json:"updated_on"`
}

// AvailabilityPublic is what other users see (with distance bucket, not coordinates)
type AvailabilityPublic struct {
	ID                  string         `json:"id"`
	UserID              string         `json:"user_id"`
	Status              string         `json:"status"`
	StartTime           time.Time      `json:"start_time"`
	EndTime             time.Time      `json:"end_time"`
	Distance            DistanceBucket `json:"distance,omitempty"`
	City                string         `json:"city,omitempty"`
	HangoutType         HangoutType    `json:"hangout_type"`
	ActivityDescription *string        `json:"activity_description,omitempty"`
	ActivityVenue       *string        `json:"activity_venue,omitempty"`
	InterestID          *string        `json:"interest_id,omitempty"`
	MaxPeople           int            `json:"max_people"`
	Note                *string        `json:"note,omitempty"`
	// User info
	UserProfile *PublicProfile `json:"user_profile,omitempty"`
	// Interest details if mutual_interest type
	Interest *Interest `json:"interest,omitempty"`
}

// HangoutRequest represents a request to join someone's availability
type HangoutRequest struct {
	ID             string     `json:"id"`
	AvailabilityID string     `json:"availability_id"`
	RequesterID    string     `json:"requester_id"`
	Note           string     `json:"note"`   // Required, min 20 chars
	Status         string     `json:"status"` // pending, accepted, declined, cancelled
	RespondedOn    *time.Time `json:"responded_on,omitempty"`
	CreatedOn      time.Time  `json:"created_on"`
}

// Hangout represents a confirmed meetup between users
type Hangout struct {
	ID                  string           `json:"id"`
	Participants        []string         `json:"participants"` // User IDs
	AvailabilityID      *string          `json:"availability_id,omitempty"`
	HangoutType         HangoutType      `json:"hangout_type"`
	ActivityDescription *string          `json:"activity_description,omitempty"`
	ScheduledTime       time.Time        `json:"scheduled_time"`
	Location            *HangoutLocation `json:"-"` // Internal only
	IsSupportSession    bool             `json:"is_support_session"`
	Status              string           `json:"status"` // scheduled, completed, cancelled, no_show
	CreatedOn           time.Time        `json:"created_on"`
	UpdatedOn           time.Time        `json:"updated_on"`
}

// HangoutStatus constants
const (
	HangoutStatusScheduled = "scheduled"
	HangoutStatusCompleted = "completed"
	HangoutStatusCancelled = "cancelled"
	HangoutStatusNoShow    = "no_show"
)

// RequestStatus constants
const (
	RequestStatusPending   = "pending"
	RequestStatusAccepted  = "accepted"
	RequestStatusDeclined  = "declined"
	RequestStatusCancelled = "cancelled"
)

// Availability constraints
const (
	MinNoteLength          = 20 // Minimum note length for friction
	MinHangoutNoteLength   = 20 // Alias for clarity
	MaxNoteLength          = 500
	MaxActivityDescLength  = 200
	MaxVenueLength         = 200
	DefaultAvailabilityTTL = 24 * time.Hour // Expire after 24 hours
)

// HangoutRequestStatus constants
const (
	HangoutRequestStatusPending   = "pending"
	HangoutRequestStatusAccepted  = "accepted"
	HangoutRequestStatusDeclined  = "declined"
	HangoutRequestStatusCancelled = "cancelled"
)

// AvailabilityLocationRequest is used when creating availability with location/radius
type AvailabilityLocationRequest struct {
	Lat    float64 `json:"lat"`
	Lng    float64 `json:"lng"`
	Radius float64 `json:"radius,omitempty"` // Default: 5km
}

// CreateAvailabilityRequest represents a request to post availability
type CreateAvailabilityRequest struct {
	Status              string                       `json:"status,omitempty"` // Default: available
	StartTime           string                       `json:"start_time"`
	EndTime             string                       `json:"end_time"`
	Location            *AvailabilityLocationRequest `json:"location,omitempty"`
	HangoutType         string                       `json:"hangout_type"`
	ActivityDescription *string                      `json:"activity_description,omitempty"`
	ActivityVenue       *string                      `json:"activity_venue,omitempty"`
	InterestID          *string                      `json:"interest_id,omitempty"`
	MaxPeople           *int                         `json:"max_people,omitempty"` // Default: 1
	Note                *string                      `json:"note,omitempty"`
	Visibility          *string                      `json:"visibility,omitempty"` // Default: circles
}

// UpdateAvailabilityRequest represents a request to update availability
type UpdateAvailabilityRequest struct {
	Status              *string `json:"status,omitempty"`
	EndTime             *string `json:"end_time,omitempty"`
	ActivityDescription *string `json:"activity_description,omitempty"`
	ActivityVenue       *string `json:"activity_venue,omitempty"`
	MaxPeople           *int    `json:"max_people,omitempty"`
	Note                *string `json:"note,omitempty"`
}

// CreateHangoutRequestRequest represents a request to join someone's availability
type CreateHangoutRequestRequest struct {
	AvailabilityID string `json:"availability_id"`
	Note           string `json:"note"` // Required, min 20 chars
}

// RespondToRequestRequest represents accepting/declining a hangout request
type RespondToRequestRequest struct {
	Status      string  `json:"status"`                // accepted, declined
	Alternative *string `json:"alternative,omitempty"` // Suggest alternative if declining
}

// ActivitySuggestion represents a suggested concrete activity
type ActivitySuggestion struct {
	Category    string `json:"category"` // nearby_now, popular, your_interests
	Icon        string `json:"icon"`
	Title       string `json:"title"`
	Description string `json:"description,omitempty"`
	Venue       string `json:"venue,omitempty"`
	Time        string `json:"time,omitempty"` // "8pm", "open till 11"
}

// AvailabilityFilter for searching available people
type AvailabilityFilter struct {
	HangoutTypes []HangoutType `json:"hangout_types,omitempty"`
	RadiusKm     *float64      `json:"radius_km,omitempty"`
	StartAfter   *time.Time    `json:"start_after,omitempty"`
	EndBefore    *time.Time    `json:"end_before,omitempty"`
	InterestID   *string       `json:"interest_id,omitempty"`
}
